threads.go

```diff
@@ -0,0 +1,384 @@
+
+// Logs returns info about the logs in the given thread.
+func (t *threads) Logs(id thread.ID) []thread.LogInfo {
+	lgs := make([]thread.LogInfo, 0)
+	for _, lid := range t.ThreadInfo(id).Logs {
+		lg := t.LogInfo(id, lid)
+		lgs = append(lgs, lg)
+	}
+	return lgs
+}
+
+// Delete the given thread.
+func (t *threads) Delete(ctx context.Context, id thread.ID) error {
+	panic("implement me")
+}
+
+// getPrivKey returns the host's private key.
+func (t *threads) getPrivKey() ic.PrivKey {
+	return t.host.Peerstore().PrivKey(t.host.ID())
+}
+
+// createLog call util.CreateLog.
+func (t *threads) createLog() (info thread.LogInfo, err error) {
+	return util.CreateLog(t.host.ID())
+}
+
+// getOrCreateLog returns the log with the given thread and log id
+// If no log exists, a new one is created under the given thread.
+func (t *threads) getOrCreateLog(id thread.ID, lid peer.ID) (info thread.LogInfo, err error) {
+	info = t.LogInfo(id, lid)
+	if info.PubKey != nil {
+		return
+	}
+	info, err = t.createLog()
+	if err != nil {
+		return
+	}
+	err = t.AddLog(id, info)
+	return
+}
+
+// getOwnLoad returns the log owned by the host under the given thread.
+func (t *threads) getOwnLog(id thread.ID) (info thread.LogInfo) {
+	for _, lid := range t.LogsWithKeys(id) {
+		if t.PrivKey(id, lid) != nil {
+			info = t.LogInfo(id, lid)
+			return
+		}
+	}
+	return
+}
+
+// getOrCreateOwnLoad returns the log owned by the host under the given thread.
+// If no log exists, a new one is created under the given thread.
+func (t *threads) getOrCreateOwnLog(id thread.ID) (info thread.LogInfo, err error) {
+	info = t.getOwnLog(id)
+	if info.PubKey != nil {
+		return
+	}
+	info, err = t.createLog()
+	if err != nil {
+		return
+	}
+	err = t.AddLog(id, info)
+	return
+}
+
+// createRecord creates a new record with the given body as a new event body.
+func (t *threads) createRecord(ctx context.Context, body format.Node, lg thread.LogInfo, settings *tserv.AddSettings) (thread.Record, error) {
+	if settings.Key == nil {
+		var key []byte
+		logKey := t.ReadKey(settings.Thread, settings.KeyLog)
+		if logKey != nil {
+			key = logKey
+		} else {
+			key = lg.ReadKey
+		}
+		var err error
+		settings.Key, err = crypto.ParseEncryptionKey(key)
+		if err != nil {
+			return nil, err
+		}
+	}
+	event, err := cbor.NewEvent(ctx, t.dagService, body, settings)
+	if err != nil {
+		return nil, err
+	}
+
+	prev := cid.Undef
+	if len(lg.Heads) != 0 {
+		prev = lg.Heads[0]
+	}
+	fk, err := crypto.ParseEncryptionKey(lg.FollowKey)
+	if err != nil {
+		return nil, err
+	}
+	rec, err := cbor.NewRecord(ctx, t.dagService, event, prev, lg.PrivKey, fk)
+	if err != nil {
+		return nil, err
+	}
+
+	t.SetHead(settings.Thread, lg.ID, rec.Cid())
+
+	return rec, nil
+}
+
+// pullLocal returns local records from the given thread that are ahead of offset but not farther then limit.
+func (t *threads) pullLocal(ctx context.Context, id thread.ID, lid peer.ID, offset cid.Cid, limit int) ([]thread.Record, error) {
```

Comment:
Check `limit` for `<0` values, and maybe put a cap in its max value to avoid massive pulls (maybe not erroring, but `max(maxPullSize, limit)` or similar (and check from the puller side that this case might happen (less than the requested limit)).

Reply:
I simplified the logic in the method a little. We have an upper limit, which when reached just returns all found values. I added a note: 

```
// It is possible to reach limit before offset, meaning that the caller
// will be responsible for the remaining traversal.
```

The same method is used on the receiver side.

Reply:
As per old logic, the default limit set in textile-core is -1. I have a change to the pull API in a different branch that removes the need for pull options. So, just leaving that default for now... the called needs to set the limit > 0(as in the tests). Otherwise, they just get no results. (-1 used to mean "all")